dlk/python/dlk/templates/include/func/concat_on_depth.h

```diff
@@ -16,90 +16,153 @@ limitations under the License.
 #ifndef DLK_FUNC_CONCAT_ON_DEPTH_H_INCLUDED
 #define DLK_FUNC_CONCAT_ON_DEPTH_H_INCLUDED
 
+#include <tuple>
+#include <type_traits>
+
 #include "global.h"
+#include "tensor_view.h"
 #include "time_measurement.h"
 
-template<class T>
-void func_ConcatOnDepth(const TensorView<T, MemoryLayout::NHWC> inputs[],
-    T_UINT n_inputs, const TensorView<T, MemoryLayout::NHWC>& output) {
-  Measurement::Start("func_ConcatOnDepth");
-  const auto shape = output.get_shape();
-  T_UINT out_height = shape[1];
-  T_UINT out_width = shape[2];
+namespace dlk {
+namespace impl {
 
-  T_UINT depths[32];
-  for (T_UINT n = 0; n < n_inputs; ++n) {
-    depths[n] = inputs[n].get_shape()[3];
+constexpr std::size_t index_channels_high(const MemoryLayout layout) {
+  switch (layout) {
+    case MemoryLayout::ChHWBCl: return 0;
+    case MemoryLayout::HWChBCl: return 2;
+    case MemoryLayout::NHWC: return 3;
+    default: return 0;
   }
-
-  T_UINT index = 0;
-  for(T_UINT n = 0; n < n_inputs; n++)
-    for(T_UINT d = 0; d < depths[n]; d++) {
-      for(T_UINT h = 0; h < out_height; h++)
-        for(T_UINT w = 0; w < out_width; w++) {
-          output(0, h, w, index) = inputs[n](0, h, w, d);
-        }
-      ++index;
-    }
-
-  Measurement::Stop();
 }
 
-template<class T>
-void func_ConcatOnDepth(const TensorView<T, MemoryLayout::HWChBCl> inputs[],
-    T_UINT n_inputs, const TensorView<T, MemoryLayout::HWChBCl>& output) {
-  Measurement::Start("func_ConcatOnDepth");
-  const auto shape = output.get_shape();
-  T_UINT out_height = shape[0];
-  T_UINT out_width = shape[1];
-  T_UINT bits = shape[3];
-
-  T_UINT depths[32];
-  for (T_UINT n = 0; n < n_inputs; ++n) {
-    depths[n] = inputs[n].get_shape()[2];
+constexpr std::size_t index_height(const MemoryLayout layout) {
+  switch (layout) {
+    case MemoryLayout::ChHWBCl: return 1;
+    case MemoryLayout::HWChBCl: return 0;
+    case MemoryLayout::NHWC: return 1;
+    default: return 0;
   }
+}
 
-  T_UINT index = 0;
-  for(T_UINT n = 0; n < n_inputs; n++)
-    for(T_UINT d = 0; d < depths[n]; d++) {
-      for(T_UINT h = 0; h < out_height; h++)
-        for(T_UINT w = 0; w < out_width; w++) {
-          for(T_UINT digit = 0; digit < bits; ++digit) {
-            output(h, w, index, digit, 0) = inputs[n](h, w, d, digit, 0);
-          }
-        }
-      ++index;
-    }
+constexpr std::size_t index_width(const MemoryLayout layout) {
+  switch (layout) {
+    case MemoryLayout::ChHWBCl: return 2;
+    case MemoryLayout::HWChBCl: return 1;
+    case MemoryLayout::NHWC: return 2;
+    default: return 0;
+  }
+}
 
-  Measurement::Stop();
+template<class T>
+T& access_ChHWBCl(const TensorView<T, MemoryLayout::ChHWBCl>& tensor,
```

Comment:
This function looks existing to access quantized tensor element, but the actual memory layout depends on the type  of the argument `tensor`. How about changing the function name to `access_quantized_tensor_elem` or something?

Just `access` or `access_tensor_element` might be enough.

Reply:
OK, fixed.